File: src/platforms/evdev/fd_store.cpp
/*
 * Copyright © Canonical Ltd.
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License version 2 or 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "fd_store.h"

#include <mir/log.h>

#include <algorithm>
#include <boost/throw_exception.hpp>

#include <stdexcept>

namespace mie = mir::input::evdev;

void mie::FdStore::store_fd(char const* path, mir::Fd&& fd)
{
    fds[path] = std::move(fd);
}

mir::Fd mie::FdStore::take_fd(char const* path)
{
    try
    {
        return fds.at(path);
    }
    catch (std::out_of_range const&)
    {
        auto it = suspended.find(path);
        if (it != suspended.end())
        {
            mir::log_warning("Requested fd for path %s was suspended", path);
            fds.insert(suspended.extract(it));
            return fds.at(path);
        }
        mir::log_warning("Failed to find requested fd for path %s", path);
    }
    return mir::Fd{};
}

void mie::FdStore::remove_fd(int fd)
{
    auto element = std::find_if(
        fds.begin(),
        fds.end(),
        [fd](auto const& pair)
        {
            return pair.second == fd;
        });

    if (element == fds.end())
    {
        mir::log_warning("Attempted to remove unmanaged fd %i", fd);
    }
    else
    {
        // Keep the suspended fd so libinput can request it again on resume.
        // Overwrite any stale entry at the same path so a double-suspend always
        // leaves the newest fd in the cache rather than silently dropping it.
        suspended[element->first] = std::move(element->second);
        fds.erase(element);
    }
}

void mie::FdStore::purge_fd(char const* path)
{
    // Remove from both maps: the device is permanently gone and its fd
    // must not be returned to libinput under any circumstances.
    fds.erase(path);
    suspended.erase(path);
}

File: src/platforms/evdev-rs/src/fd_store.rs
/*
 * Copyright © Canonical Ltd.
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License version 2 or 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use std::collections::HashMap;
use std::os::fd::OwnedFd;

/// Manages pre-acquired file descriptors for device nodes.
///
/// When a device is activated via ConsoleServices, its fd is stored here
/// so that libinput's `open_restricted` callback can claim it without
/// blocking. The fd is never removed on claim — each `open_restricted`
/// call receives a fresh `dup` — so libinput can re-open the device as
/// many times as it likes (e.g. after a lid-switch event) without any
/// separate backup mechanism. The fd is only released when the device is
/// explicitly removed via `remove`.
pub struct FdStore {
    fds: HashMap<String, OwnedFd>,
}

impl FdStore {
    pub fn new() -> Self {
        FdStore {
            fds: HashMap::new(),
        }
    }

    /// Store a pre-acquired file descriptor for a device path.
    /// Any previous fd for the same device is closed automatically.
    pub fn store(&mut self, devnode: &str, fd: OwnedFd) {
        self.fds.insert(devnode.to_string(), fd);
    }

    /// Return a dup'd fd for a device path without consuming the stored fd.
    /// Returns `None` if no fd is stored for this device.
    /// The caller owns the returned fd and is responsible for closing it.
    pub fn claim(&self, devnode: &str) -> Option<OwnedFd> {
        self.fds.get(devnode)?.try_clone().ok()
    }

    /// Remove and close the stored fd for a device (called on device removal).
    pub fn remove(&mut self, devnode: &str) {
        self.fds.remove(devnode);
    }
}
